pandas/core/arrays/datetimes.py

```diff
@@ -477,10 +477,20 @@ def _generate_range(
             arr = arr.astype("M8[ns]", copy=False)
             index = cls._simple_new(arr, freq=None, dtype=dtype)
 
-        if not left_closed and len(index) and index[0] == start:
-            index = index[1:]
-        if not right_closed and len(index) and index[-1] == end:
-            index = index[:-1]
+        # do not remove when one side is inclusive
+        # and removing would leave index empty
+        to_remove_any = not (
+            (left_inclusive or right_inclusive)
+            and len(index)
+            and start == index[0]
+            and start == end
+        )
+
+        if to_remove_any:
+            if (not left_inclusive) and len(index) and index[0] == start:
```

Comment:
you are reusing the same logic tests deriving `to_remove_any`. I haven't come up with a better way but instinctively this feels a complicated logic path. Have you tried refactoring to a simpler setup?

Reply:
I refactored the logic a little and I am not sure if it is any better though